lib/Sema/CSDiag.cpp

```diff
@@ -6508,7 +6554,75 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
         }
         return true;
       }
-      
+
+      if (inferredArgCount == 1 && actualArgCount > 1) {
+        // Let's see if inferred argument is actually a tuple inside of Paren.
+        if (auto *argTupleTy = inferredArgType->getAs<TupleType>()) {
+          // Looks like the number of closure parameters matches number
+          // of inferred arguments, which means we can warn about SE-0110
+          // and provide a proper fix-it.
+          if (argTupleTy->getNumElements() == actualArgCount) {
+            auto diag = diagnose(params->getStartLoc(),
+                                 diag::closure_parameter_tuple_destructuring,
+                                 argTupleTy);
+
+            Type actualArgType;
+            if (auto *actualFnType = CE->getType()->getAs<AnyFunctionType>())
+              actualArgType = actualFnType->getInput();
+
+            auto typeComparator = [&](Type lhs, Type rhs) -> bool {
+              auto *typeVar = lhs->getAs<TypeVariableType>();
+              if (!typeVar)
+                return lhs->isEqual(rhs);
+
+              // Type variables are allowed only in the actual arguments.
+              if (rhs->getKind() == TypeKind::TypeVariable)
+                return false;
+
+              // If there is an inferred type associated with closure
+              // parameter, let's check that against inferred parameter type.
+              if (auto fixedType = CS->getFixedType(typeVar))
+                return fixedType && fixedType->isEqual(rhs);
+
+              // type variabls in this case represent types of
+              // the closure parameters and we only check structure
+              // of the type itself.
+              return true;
+            };
+
+            // Can't produce fix-it, because inferred parameter
+            // and actual argument type structure are not the same.
+            if (!matchArgStructure(actualArgType, argTupleTy, typeComparator))
```

Comment:
I'm pretty sure you can just drop this. Fix-its on notes are allowed to be wrong sometimes, but even then I can't think of a situation where it would be.